### PR TITLE
[enterprise-4.11] [OSDOCS-4789]: Machine deletion hooks

### PR DESCRIPTION
**Pull request by jeana-redhat**
Cherrypicked from 7962d98ff32c9f4d0575bbf777112d51089d3b3d xref: #55175

Selective manual cherrypick because 4.11 has quorum protection but no control plane machine set.

@JoelSpeed and @miyadav PTAL :pray: 

Link to docs preview:
* [Deleting a specific machine](https://59032--docspreview.netlify.app/openshift-enterprise/latest/machine_management/deleting-machine.html#machine-delete_deleting-machine)
* [Lifecycle hooks for the machine deletion phase](https://59032--docspreview.netlify.app/openshift-enterprise/latest/machine_management/deleting-machine.html#machine-lifecycle-hook-deletion_deleting-machine)
* [Scaling a compute machine set manually](https://59032--docspreview.netlify.app/openshift-enterprise/latest/machine_management/manually-scaling-machineset.html#machineset-manually-scaling_manually-scaling-machineset) (for [OCPBUGS-6257](https://issues.redhat.com/browse/OCPBUGS-6257), this bit needs additional backporting through 4.9)
* [Replacing an unhealthy etcd member](https://59032--docspreview.netlify.app/openshift-enterprise/latest/backup_and_restore/control_plane_backup_and_restore/replacing-unhealthy-etcd-member.html) (+Additional resources) 
* Operators overview > [For developers](https://59032--docspreview.netlify.app/openshift-enterprise/latest/operators/index.html#operators-overview-developer-tasks) (+Additional resources)

### CODE REVIEW

**Comment by jeana-redhat** [comment]
Oops I seem to have done something bad to this with git :grimacing:

File: modules/machine-lifecycle-hook-deletion-etcd.adoc [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+. The machine controller removes the instance from the infrastructure provider.
+. The machine controller deletes the `Node` object.
+
+////
+//Is there something helpful to add here for OCP 4.11?
+.YAML snippet demonstrating the etcd quorum protection `preDrain` lifecycle hook
+[source,yaml]
+----
+apiVersion: machine.openshift.io/v1beta1
+kind: ControlPlaneMachineSet
```
Comment (JoelSpeed) [inline modules/machine-lifecycle-hook-deletion-etcd.adoc:47]: This should have been `Machine` not `ControlPlaneMachineSet`, it's still relevant in 4.11

File: modules/machine-lifecycle-hook-deletion-format.adoc [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+[id="machine-lifecycle-hook-deletion-format_{context}"]
+= Deletion lifecycle hook configuration
+
+The following YAML snippets demonstrate the format and placement of deletion lifecycle hook configurations within a machine set:
+
+.YAML snippet demonstrating a `preDrain` lifecycle hook
+[source,yaml]
+----
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineSet
```
Comment (JoelSpeed) [inline modules/machine-lifecycle-hook-deletion-format.adoc:15]: Probably better to be on the Machine rather than MachineSet for all the examples

**Review by JoelSpeed — COMMENTED** [review]
/lgtm

**Comment by kalexand-rh** [comment]
@miyadav, will you please approve this change?

File: modules/machine-lifecycle-hook-deletion.adoc [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+** The machine gets a `machine.openshift.io/delete-machine` annotation.
+** The machine set that manages the machine marks it for deletion to reduce the replica count as part of reconciliation.
+** The cluster autoscaler identifies a node that is unnecessary to meet the deployment needs of the cluster.
+** A machine health check is configured to replace an unhealthy machine.
+. The machine enters the `Deleting` phase, in which it is marked for deletion but is still present in the API.
+. If a `preDrain` lifecycle hook exists, the hook-implementing controller that manages it does a specified action.
++
+Until all `preDrain` lifecycle hooks are satisfied, the machine status condition `Drainable` is set to `False`.
+. There are no unresolved `preDrain` lifecycle hooks and the machine status condition `Drainable` is set to `True`.
+. The machine controller attempts to drain the node that is backed by the machine.
```
Comment (adellape) [inline modules/machine-lifecycle-hook-deletion.adoc:69]: Same wonder re: colon or not, leading into sub-bullets. ... This is probably fine/correct.

**Comment by jeana-redhat** [comment]
@adellape thank you! Given that this is a (somewhat involved) cherrypick, could I merge this and pick up your fixes in a separate PR? Not sure what the best way is to make the least things break :sweat_smile:

**Pull request merged**